### PR TITLE
New `CentralThreadlyPool`

### PR DESCRIPTION
**Pull request by jentfoo**
This is incomplete, still requiring unit tests as well as an expansion in functionality.
This adds a central (static) pool, which allows you to request other pools based off an expected task specification.  This central class will manage its internal pools to try and provide the best experience for the implementer, while not requiring pool lifecycle management (like shutdown), and minimizing thread sprawl.

In order to accomplish this easily `ExecutorLimiter` was updated to allow it's limit to be adjusted post-construction.  And `PriorityScheduler` now allows the pool size to be adjusted by a delta rather than requiring an absolute value.

Disadvantages in using a class like this:
* Overhead, though attempted to kept minor, does exist.  Primarily in additional GC of structures, and the need to use a queue limiter to ensure that guarded threads are always available to those who need them.

### CODE REVIEW

File: src/main/java/org/threadly/concurrent/CentralThreadlyPool.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    private final int personallyGarunteedThreads;
+    private final int maxThreads;
+    
+    public GuaranteedConnectionProtectingScheduler(TaskPriority priority, 
+                                                  int personallyGarunteedThreads, int maxThreads) {
+      super(masterScheduler(priority), 1);
+      
+      this.gcReference = personallyGarunteedThreads > 0 ? new PoolResizer(personallyGarunteedThreads) : null;
+      this.personallyGarunteedThreads = personallyGarunteedThreads > 0 ? personallyGarunteedThreads : 0;
+      this.maxThreads = maxThreads < 1 ? Integer.MAX_VALUE : maxThreads;
```
Comment (lwahlmeier) [inline src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: I feel like saying I need "X" max threads should make the thread pool at least that size, and if they say N dedicated the pool should be at least X with N not being shared on another pool.  Basically whatever the most threads requested for a pool is is how many threads should be in the global pool at a minimum, plus any dedicated ones or any other pools

so in a case of these sub pool requests:
max    dedicated
10         5
5           1
20         3
50         1

there would be 60 threads, 50 for general processing and 10 dedicated.  if the last pool GC's we would adjust to have 29 threads total, 20 shared and 9 dedicated.

File: src/main/java/org/threadly/concurrent/CentralThreadlyPool.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    protected void finalize() {
+      GUARANTEED_THREADS.add(-amount);
+      MASTER_SCHEDULER.adjustPoolSize(-amount);
+    }
+  }
+
+  // TODO - this is slightly flawed, we need to have a way for the pool to grow without 
+  //          requiring protected threads otherwise we will never use available resources.
+  //          If it does not make sense changing when the pool is expanded (or how), then this 
+  //          class is probably providing a protection that makes everything useless...quite a thinker
```
Comment (lwahlmeier) [inline src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: Yeah, The only other way I can think to watch this is for threads to have some kind of processing timer we check, or have a recurring task we watch for getting blocked, both have different issues, mainly they could grow like crazy if load is high even in situations where threads are not stuck in blocking operations they would just grow different in different situations.

File: src/main/java/org/threadly/concurrent/CentralThreadlyPool.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+ * 
+ * @since 5.4
+ */
+public class CentralThreadlyPool {
+  protected static final int LOW_PRIORITY_MAX_WAIT_IN_MS = 1000;  // TODO - configurable?  Diff from default?
+  protected static final PriorityScheduler MASTER_SCHEDULER;
+  protected static final PrioritySchedulerService LOW_PRIORITY_MASTER_SCHEDULER;
+  protected static final SchedulerService COMPUTATION_POOL;
+  protected static final SchedulerService LOW_PRIORITY_POOL;
+  protected static final PrioritySchedulerService SINGLE_THREADED_LOW_PRIORITY_POOL;
```
Comment (lwahlmeier) [inline src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: I personally would also like to see a centeral Keyed scheduler/executor.  I would use that alot honestly.

**Comment by lwahlmeier** [comment]
mostly looks good to me, I put comments in places I had some thoughts/ideas.

File: src/main/java/org/threadly/concurrent/CentralThreadlyPool.java (hunk 3 above)
Comment (jiming) [reply src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: Agree. That is one important feature of CTP I expected.

Comment (jentfoo) [reply src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: I considered having one, but the single thread scheduler design kind of overtook it.  Let me know your thoughts on if it would be a better option?  What would the advantage be of a global keyed executor?

Comment (jiming) [reply src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: I think the benefits of keyed are:

- Config or Show infomation of Executors managed by CTP. Manageable is very important to online system. I'd really like to change some config without restart server. Of course, check CTP information is important as well.
- I might wanted to temproarily increase priority of Executor by its name

Comment (jentfoo) [reply src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: @jiming I talked more about this with @lwahlmeier directly.  I think part of the problem is we are trying to solve different things (and that's in part what makes this so hard).

@lwahlmeier was thinking that this could become a replacement for all pools within a service.  And while I guess that's technically possible, that's not what I was thinking.

I was thinking this really would be useful for replacing small one off pools.  Things like "I need a cleanup task for this class", or "I want to report metrics / something".  I would think things like request pools, or pools which are heavily active / important to the service, would not be replaced by this.  I personally would not want to give up the control, or the performance sacrifices in replacing such important pools with something like this.

What were you thinking this role would be?  Did you want it to replace all pooling, or just common / small, utility pools?

Comment (jiming) [reply src/main/java/org/threadly/concurrent/CentralThreadlyPool.java]: @jentfoo 

Now I can understand your means. 

Maybe next version, you could add some more features. My options is that you already go this far. Why not go a little further? Make threadly a de facto for ThreadPool Manager in JDK.

> What were you thinking this role would be? Did you want it to replace all pooling, or just common / small, utility pools?

I like a ONE who rule them all Thread Pool manager?

Thanks!

**Comment by jentfoo** [comment]
@lwahlmeier and @jiming I am for sure going to make this first iteration on this focused around auxilary and low-performance impact pool replacements.  Not designed to replace needs where `Keyed` executors might be needed.

That said, I just rebased on the current master and pushed another potential implementation to this.  It's very similar to how it was before, with the primary change coming in how we represent the "ranged" pools.

In general this works by providing the following:
* A "computation" pool which will have a thread count matching the cores on the system.  These protected threads are to be used by threads which are doing in-thread computational work.  This returned instance is always shared.
* A "low priority" pool, which may be constructed to only one thread, or use all available shared threads.  This is a pool where execution time is not important, and allows a central place for low priority cleanup operatons.  This returned instance is always shared.
* The ability to produce a single threaded pool.  This pool will execute the tasks submitted to a given returned instance in a single threaded way.
* A ranged pool.  Specifying a number of minimum threads to spawn for it, as well as a potential limit on shared threads that it can consume.  By default there is only one shared thread (regardless of what is provided here).  In order to increase potential shared threads `increaseGeneralProcessingThreads(int)` must be invoked.

So in short ranged pools will create their minimum needed threads and those will be protected.  Regardless how big or small the maximum threads specified initially they will all compete for the `1` shared thread.  Unless an increase is invoked.

I have been struggling to find a behavior pattern where the shared thread size is dynamic, or even based off the max threads supplied (suggestions welcome).

**Comment by jentfoo** [comment]
@jiming and @lwahlmeier 

I have not forgotten about this.  I have given it some more thought, but want your opinion on something.  No need to refresh with the previous details in the PR, I will explain it all here.

I suspect most people will (and should) use `.computationPool()`, `.lowPriorityPool(boolean)`, `.singleThreadPool()`, and `threadPool(int)`.  These functions have pretty safe semantics around them.  In all cases they are guaranteed to have the threads requested available for execution.  In this case `CentralThreadlyPools` primary job is to just encourage reuse across the pools and reduce thread churn as much as possible.

In the current state of this PR there is an additional concept that exists but will only be available to a small subset of users.  I am wondering if we should continue to maintain it or remove it for simplicity and just instead focus on the above implementations.

This concept is "general processing threads"  By default there will be one, but they can be increased by invoking `increaseGeneralProcessingThreads(int)` (they can only be increased, never removed).  These threads are used by the `lowPriorityPool`, and if you request `rangedThreadCountPool` with either zero guaranteed threads, or a max threads which is higher than the guaranteed threads (or -1 to allow use of all general processing threads).

These general processing threads could in theory provide hints to further reduce thread churn.

However they complicate the API.  I worry about potential danger with this potentially increasing common pooling of threads.  And there is a minor performance hit in order to be able to dynamically size our limiter for when they are added.

Do you guys think these general processing threads as a concept makes sense, or should this class for now just focus on removing the pool lifecycle and reducing thread churn as much as possible with those simple semantics?

Beyond that the only thing left here is probably the potential to rename threads, as well as unit tests.  So this is probably going to get released in 5.7 for sure.

**Comment by jiming** [comment]
@jentfoo 

Thank you!

I am sorry I am on an urgent project now and have no time in a couple of weeks. I will sure checkk when I am freed.

BTW, what do you think to totally support java 8, using `CompletableFuture` replace `ListenableFuture` you defined here. I suppose supporting java 8 and above will reduce a lot of code and interfaces here.

And I'd like to recommend a great lib to record metrics. https://github.com/micrometer-metrics/micrometer. Spring and Spring boot is going to use it as well. 

Have a nice day!

Jiming

**Comment by jentfoo** [comment]
> BTW, what do you think to totally support java 8, using `CompletableFuture` replace `ListenableFuture` you defined here.

A resounding "maybe".  There is some things I like about this interface, but lots I don't like as well.  Though it is interesting how they put most of the functionality we have in `FutureUtils` into the interface itself.  The reason I would see to do it is not reduced code but for improved compatibility with other libraries and the rest of the application.

In any case this would be a change we really will want to think about till we start thinking about 6.0.  If this is something you are interested, please open an issue so we don't forget to consider it for 6.0

Take a look when you get a chance.  The only big decision we have before we push this forward is "do we want to keep this concept of "general processing threads".  They complicate the API, and could potentially be dangerous if abused heavily.  But they also have the potential to further reduce thread churn in specific applications.

We could always remove them for now and add them back later as well.

**Comment by jiming** [comment]
Make a new issue :)

**Comment by jentfoo** [comment]
@lwahlmeier I think I got this to a point where I am personally happy with it.  I don't like the API bloat for thread renaming, but I do think it's an important feature.

Would you mind looking and giving me your thoughts?

@jiming I would like your thoughts as well if you have time, but also would like to include this in 5.7 now that it has been in progress so long.

**Comment by jentfoo** [comment]
Ok, I added a few more changes after making use of it internally to remove our Watchdog pool and Clock thread.  But I am for sure done now (until I get feedback at least) :)

**Comment by jentfoo** [comment]
I talked with @lwahlmeier and we are good to go as a first iteration of this.  Additional changes or improvements can always be incorporated later.  So feel free to open issues with suggestions.
